File: tests/compiled_workloads/adapters/peer/cross-platform-notes.mjs
// #1414 peer adapter. Existing authorized web incumbent:
// TypeScript/ECMAScript, commit 5be33469d551655d878876faa9e30aa3b49f8ee9.
// The implementation below is independent: the revision identifies the
// selected web boundary, not copied adapter source.

const ACTIONS = new Set([
  "add",
  "edit",
  "search",
  "save",
  "reload",
  "readonly",
  "corrupt",
]);
const FIELDS = new Set(["title", "body", "query"]);

function newModel() {
  return {
    notes: [],
    active: -1,
    focus: "title",
    query: "",
    search: "not-run",
    persistence: "not-saved",
    readonly: "not-run",
    corrupt: "not-run",
    readOnlyStorage: false,
    savedNotes: null,
    unknown: false,
  };
}

function cloneNotes(notes) {
  return notes.map((note) => ({ title: note.title, body: note.body }));
}

function parseLine(line) {
  const separator = line.indexOf(":");
  if (separator < 0) return { kind: "unknown" };

  const name = line.slice(0, separator);
  const value = line.slice(separator + 1);
  if (name === "key" && ACTIONS.has(value)) {
    return { kind: "action", name: value };
  }
  if (FIELDS.has(name)) return { kind: "field", name, value };
  return { kind: "unknown" };
}

function updateSearch(model) {
  model.search = model.notes.some(
    (note) =>
      note.title.includes(model.query) || note.body.includes(model.query),
  )
    ? "found"
    : "not-found";
}

function applyAction(model, name) {
  switch (name) {
    case "add":
      model.notes.push({ title: "", body: "" });
      model.active = model.notes.length - 1;
      model.focus = "title";
      break;
    case "edit":
      if (model.notes.length > 0) model.active = model.notes.length - 1;
      model.focus = "title";
      break;
    case "search":
      updateSearch(model);
      break;
    case "save":
      if (!model.readOnlyStorage) {
        model.savedNotes = cloneNotes(model.notes);
        model.persistence = "saved";
      }
      break;
    case "reload":
      if (model.savedNotes !== null) {
        model.notes = cloneNotes(model.savedNotes);
        model.active = model.notes.length - 1;
        model.persistence = "reloaded";
      }
      break;
    case "readonly":
      model.readOnlyStorage = true;
      model.readonly = "blocked";
      break;
    case "corrupt":
      model.corrupt = "rejected";
      break;
    default:
      model.unknown = true;
  }
}

function applyEvent(model, event) {
  if (event.kind === "unknown") {
    model.unknown = true;
    return;
  }
  if (event.kind === "action") {
    applyAction(model, event.name);
    return;
  }

  if (event.name === "query") {
    model.query = event.value;
    if (model.search !== "not-run") updateSearch(model);
    return;
  }
  if (model.active >= 0) {
    model.notes[model.active][event.name] = event.value;
    if (model.search !== "not-run") updateSearch(model);
  }
}

export function renderNotes(raw) {
  const model = newModel();
  for (const line of String(raw ?? "").split(/\r?\n/)) {
    if (line.length > 0) applyEvent(model, parseLine(line));
  }

  const lines = [
    `notes=${model.notes.length}`,
    `focus=${model.focus}`,
    `search=${model.search}`,
    `persistence=${model.persistence}`,
    `readonly=${model.readonly}`,
    `corrupt=${model.corrupt}`,
  ];
  if (model.unknown) lines.push("reject=unknown-key");
  return `${lines.join("\n")}\n`;
}

const runningAsCli =
  typeof process !== "undefined" &&
  process.argv[1]?.endsWith("cross-platform-notes.mjs");

if (runningAsCli) {
  const fs = await import("node:fs");
  process.stdout.write(renderNotes(fs.readFileSync(process.argv[2], "utf8")));
} else if (typeof window !== "undefined") {
  for (const line of renderNotes(globalThis.__compiledWorkloadInput).trimEnd().split("\n")) {
    console.log(line);
  }
}

File: tests/compiled_workloads/adapters/peer/candidates/cross-platform-notes.swift
import Foundation

let arguments = CommandLine.arguments
if arguments.count != 2 { exit(64) }
guard let text = try? String(contentsOfFile: arguments[1], encoding: .utf8) else { exit(2) }
var notes = 0
var focus = "title"
var search = "not-run"
var persistence = "not-saved"
var readonly = "not-run"
var corrupt = "not-run"
var active = false
var saved = false
var readOnlyStorage = false
var unknown = false
var query = ""
var title = ""
var body = ""
func updateSearch() {
    search = title.contains(query) || body.contains(query) ? "found" : "not-found"
}
for line in text.split(whereSeparator: \.isNewline) {
    let fields = line.split(separator: ":", maxSplits: 1).map(String.init)
    guard fields.count == 2 else { unknown = true; continue }
    let key = fields[0], value = fields[1]
    if key == "key" {
        switch value {
        case "add": notes += 1; active = true; title = ""; body = ""; focus = "title"
        case "edit": if notes > 0 { active = true }; focus = "title"
        case "search": updateSearch()
        case "save": if !readOnlyStorage { saved = true; persistence = "saved" }
        case "reload": if saved { persistence = "reloaded" }
        case "readonly": readOnlyStorage = true; readonly = "blocked"
        case "corrupt": corrupt = "rejected"
        default: unknown = true
        }
    } else if key == "title" {
        if active { title = value; if search != "not-run" { updateSearch() } }
    } else if key == "body" {
        if active { body = value; if search != "not-run" { updateSearch() } }
    } else if key == "query" {
        query = value; if search != "not-run" { updateSearch() }
    } else {
        unknown = true
    }
}
print("notes=\(notes)")
print("focus=\(focus)")
print("search=\(search)")
print("persistence=\(persistence)")
print("readonly=\(readonly)")
print("corrupt=\(corrupt)")
if unknown { print("reject=unknown-key") }
